### DISCUSSION TITLE
STM32F3x DMA configuration for UART

### DISCUSSION THREAD

**Discussion by sbendt** [zephyrproject-rtos/zephyr#67303] [Q&A] [answered]
Hi,

I am trying to use the async transfers with usarts, but it fails to enable the UART (using uart_rx_enable), due to no DMA configured for the usart.

I have not been able to modify the device tree for DMA, but so far unsuccessful.

On the dma1 node, I have changed the status from "disabled" to "okay", and modified the usart3 to reference the dma1 in "dmas".

&dma1 {
	status = "okay";
};

&usart3 {
    pinctrl-0 = <&usart3_tx_pb10 &usart3_rx_pb11>;
    pinctrl-names = "default";
    current-speed = <19200>;
    dmas = <&dma1 3 1 (STM32_DMA_PERIPH_TX | STM32_DMA_PRIORITY_HIGH)>,
          <&dma2 2 1 (STM32_DMA_PERIPH_RX | STM32_DMA_PRIORITY_HIGH)>;
    //dma-names = "tx", "rx";
    status = "okay";
};
 
It fails with:
zephyr.dts:209.4-64: Warning (dmas_property): /soc/serial@40004800:dmas: cell 3 is not a phandle reference
zephyr.dts:209.4-64: Warning (dmas_property): /soc/serial@40004800:dmas: Could not get phandle node for (cell 3)

The MCU is the STM32F303CBT6 .
 
Any idea how to configure DMA for usarts correctly?

**Comment by FRASTM** [comment, marked as answer]
Hello,

When using the usart3 with DMA
the dts should look like : 
```
 &usart3 {
	...
	dmas = <&dma1 2 STM32_DMA_PERIPH_TX>,
		<&dma1 3 STM32_DMA_PERIPH_RX>;
	dma-names = "tx", "rx";
};

```
This is given by the table of the [Ref Man](https://www.st.com/resource/en/reference_manual/rm0316-stm32f303xbcde-stm32f303x68-stm32f328x8-stm32f358xc-stm32f398xe-advanced-armbased-mcus-stmicroelectronics.pdf) for DMA1 request for each channel.

<img width="530" alt="Capture" src="https://github.com/zephyrproject-rtos/zephyr/assets/51417392/7df9f21e-84ef-49d0-8b1a-4d5cbb127903">

**Reply by FRASTM** [reply]
Running the tests/drivers/uart/uart_async_api for the usart3 on the corresponding board (connecting Tx to Rx pins) should PASS

**Comment by sbendt** [comment]
Thank you for the reply and information - appreciate it!

I did the modifications and I am able to enable the UART and get some of the events in the callback function. However, when transmitting data, the application crashes and I never get the UART_RX_RDY event. I still have some debugging to do, but let me know if you have any recommendations or things to check.

**Reply by FRASTM** [reply]
did you try with `current-speed = <115200>;`

**Reply by sbendt** [reply]
I got the receive part working, but for some reason transmit crashes my application.

The error handler  k_sys_fatal_error_handler is called with reason=26 initially and then reason=4 after.

I am not sure how to troubleshoot that. 

&dma1 {
	status = "okay";
};

&usart2 {
	pinctrl-0 = <&usart2_tx_pa2 &usart2_rx_pa3>;
	pinctrl-names = "default";
	current-speed = <115200>;
	dmas = <&dma1 7 (STM32_DMA_PERIPH_TX | STM32_DMA_PRIORITY_HIGH)>,  
	       <&dma1 6  (STM32_DMA_PERIPH_RX | STM32_DMA_PRIORITY_HIGH)>;
	dma-names = "tx", "rx";
	status = "okay";
};

The channel 6 and 7 should be in accordance with the manual.

**Comment by R-Lourens** [comment]
@sbendt Hello, I am also looking to implement uart using the Zephyr async api. My board is a STM32F3 Nucleo-144. So far this discussion has helped me alot but I would like to know if you are willing to share how your code looks in your main.c file? (This is my first time working on DMA's) I have looked at [https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/uart/uart_async_api](url) and I think that I could figure out how to use the async api from there also?
